std/utils.asm

```diff
@@ -40,17 +40,24 @@ let new_bool: -> expr = constr || {
     x
 };
 
-/// Creates a new witness column which is 1 on each row where `x` is zero
-/// and 0 otherwise.
+/// Returns a 2nd degree boolean expression that is 1 if and only if `x` is not zero.
+///
+/// Introduces one new witness column and one new constraint.
+let is_not_zero: expr -> expr = constr |x| {
+    // The inverse of "x" if it is not zero, otherwise undefined.
+    let inverse;
+
+    // Ensures that "inverse * x" is either 0 or 1:
+    (inverse * x - 1) * x = 0;
```

Comment:
Is this a degree-3 constraint? Is that fine for our backends?

Reply:
The degree is 3 only if x is degree 1.

The final degree is, of course, 2*deg(x) + 1.

IIRC, 3 is the target degree we support. It works on plonky3.

Also, it seems very wrong that the user has to worry about this. It feels that the breakup of compatible polynomials should be a compilation step, possibly dependent on the constraints of the backend.

Reply:
Create an issue about this, but I admit the thoughts in there are convoluted and needs refining: #1971